### PR TITLE
[SCEV] Don't use non-deterministic constant folding for trip counts

### PR DESCRIPTION
**Pull request by nikic**
When calculating the exit count exhaustively, if any of the involved operations is non-deterministic, the exit count we compute at compile-time and the exit count at run-time may differ. Using these non-deterministic constant folding results is only correct if we actually replace all uses of the instruction with the value. SCEV (or its consumers) generally don't do this.

Handle this by adding a new AllowNonDeterministic flag to the constant folding API, and disabling it in SCEV. If non-deterministic results are not allowed, do not fold FP lib calls in general, and FP operations returning NaNs in particular. This could be made more precise (some FP libcalls like fabs are fully deterministic), but I don't think this that precise handling here is worthwhile.

Fixes the interesting part of https://github.com/llvm/llvm-project/issues/89885.

### CODE REVIEW

File: llvm/include/llvm/Analysis/ConstantFolding.h [hunk 1]
```diff
@@ -68,9 +68,15 @@ Constant *ConstantFoldConstant(const Constant *C, const DataLayout &DL,
 /// fold instructions like loads and stores, which have no constant expression
 /// form.
 ///
+/// In some cases, constant folding may return one value chosen from a set of
+/// multiple legal return values. Using such a result is usually only valid if
```
Comment (fhahn) [inline llvm/include/llvm/Analysis/ConstantFolding.h:72]: Might be worth mentioning the NaN case as example?

**Review by efriedma-quic — COMMENTED** [review]
Do we need to do something with fast-math flags?

**Comment by nikic** [comment]
> Do we need to do something with fast-math flags?

I think so. I've disabled folding for nsz and reassoc/arcp/contract. I didn't try to come up with realistic examples for them, but I think they could all cause issues.

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Pull request merged**